carthage/deployment.py

```diff
@@ -0,0 +1,494 @@
+    #identified as a Deployable, it will also appear in *failures*.
+    instantiation_failure_leaves: dict[InjectionKey, dependency_introspection.FailedInstantiation] = \
+        dataclasses.field(repr=False, default_factory=lambda: {})
+
+    def is_successful(self):
+        return self.successes and  not (self.failures or self.dependency_failures or self.instantiation_failure_leaves)
+
+    def __contains__(self, d:Deployable):
+        if d in self.successes:
+            return True
+        if any(x.deployable is d for x in self.failures):
+            return True
+        if any(x.deployable is d for x in self.dependency_failures):
+            return True
+        return False
+    
+    def _injection_failed_cb(self, target, target_key, context, **kwargs):
+        try:
+            obj = context.get_value_no_instantiate()
+        except AttributeError: obj = None
+        if hasattr(obj, 'find') and hasattr(obj, 'find_or_create'):
+            deployable = obj
+            deployment_failure = DeploymentFailure(
+                deployable=deployable,
+                exception=target.exception,
+                # Turn tuple() into None
+                dependency_path=target.dependency if target.dependency else None)
+        else:
+            deployable = None
+            deployment_failure = None
+        if not target.dependency:
+            # We take the first one if there are duplicate InjectionKeys
+            self.instantiation_failure_leaves.setdefault(target_key, target)
+            if deployment_failure and deployable not in self:
+                self.failures.append(deployment_failure)
+        else: # there is a dependency_path
+            if deployment_failure and deployable not in self:
+                self.dependency_failures.append(deployment_failure)
+
+    def method_callback(self, deployable:Deployable):
+        def callback(future):
+            if  deployable in self:
+                # We already noticed an instantiation failure
+                if not future.exception():
+                    logger.warning("successful method callback for %s, already recorded as a failure in the DeploymentResult", deployable)
+                    return
+            if exception := future.exception():
+                if isinstance(exception, InjectionFailed) and exception.dependency_path:
+                    self.dependency_failures.append(DeploymentFailure(
+                        deployable, exception=exception.__context__,
+                        dependency_path=exception.dependency_path))
+                else:
+                    self.failures.append(DeploymentFailure(
+                        deployable=deployable,
+                        exception=exception))
+            else: # no exception
+                future.result() # so it is read
+                self.successes.append(deployable)
+
+        return callback
+
+    def find_callback(self, deployable: Deployable):
+        def callback(future):
```

Comment:
I noticed the method `callback` inside both `find_callback` and `method_callback` are eerily similar. 

The only differences being how the `else` is handled as well as on line 167 you set `exception=exception.__context__` and on line 189 you set `exception=exception`. 

What is the difference between using `exception.__context__` and just `exception`?